src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicRemoveUnusedLocals.vb

```diff
@@ -57,9 +57,17 @@ Namespace Microsoft.CodeQuality.VisualBasic.Analyzers.Maintainability
 
                         operationBlockContext.RegisterOperationAction(
                         Sub(operationContext)
-                            Dim locals = DirectCast(operationContext.Operation, IVariableDeclarationGroupOperation).GetDeclaredVariables()
-                            For Each local In locals
-                                mightBecomeUnusedLocals.Add(local)
+                            Dim declarations = DirectCast(operationContext.Operation, IVariableDeclarationGroupOperation).Declarations
+
+                            For Each declaration In declarations
+                                ' Declarations with invocation expressions as initializers declare unused locals.
```

Comment:
I am a bit confused here. Why do we only exclude invocation in the initializers? There might be lot more other code patterns that can cause side effects, whose removal can cause semantic changes in code. I think we should change this analyzer to match the C# compiler behavior and only flag if (1) `.Initializer?.Value.ConstantValue.HasValue` is true and (2) For all subsequent assignments to this variable, the right hand side has a ConstantValue.

Reply:
I agree with this. This is conservative and reasonable. BTW, in this case, 'Dim a as New Integer' should be considered as use because the New initializer can have side effects.

---
In reply to: [152664718](https://github.com/dotnet/roslyn-analyzers/pull/1336#discussion_r152664718) [](ancestors = 152664718)

Reply:
> BTW, in this case, 'Dim a as New Integer' should be considered as use because the New initializer can have side effects.

Agreed. We can skip variables initializer with `AsNewClause`